src/3D/Camera.cpp

```diff
@@ -9,28 +9,136 @@
 {
 	return GetProjectionMatrix() * GetViewMatrix();
 }
 
+const std::optional<ecs::components::Transform> Camera::RaycastCamToLand()
+{
+	// get the hit by raycasting to the land down the camera ray
+	ecs::components::Transform intersectionTransform;
+	float intersectDistance = 0.0f;
+	glm::vec3 viewVec = GetForward();
+	auto& dynamicsSystem = Game::instance()->GetDynamicsSystem();
+	if (auto hit = dynamicsSystem.RayCastClosestHit(_position, viewVec, 1e10f))
+	{
+		intersectionTransform = hit->first;
+		return std::make_optional(intersectionTransform);
+	}
+	else if (glm::intersectRayPlane(_position, viewVec, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f),
+	                                intersectDistance))
+	{
+		intersectionTransform.position = _position + viewVec * intersectDistance;
+		intersectionTransform.rotation = glm::mat3(1.0f);
+		return std::make_optional(intersectionTransform);
+	}
+	// else // point 100 away from camera to rotate around etc. not currently used
+	//{
+	// intersectionTransform.position = _position + viewVec * 100.f;
+	// intersectionTransform.rotation = glm::mat3(1.0f);
+	// return std::make_optional(intersectionTransform);
+	//}
+	return std::nullopt;
+}
+
+const std::optional<ecs::components::Transform> Camera::RaycastMouseToLand()
+{
+	// get the hit by raycasting to the land down via the mouse
+	ecs::components::Transform intersectionTransform;
+	float intersectDistance = 0.0f;
+	int sWidth, sHeight;
+	Game::instance()->GetWindow()->GetSize(sWidth, sHeight);
+	glm::vec4 viewport = glm::vec4(0, 0, sWidth, sHeight);
+	glm::vec3 rayOrigin, rayDirection;
+	glm::ivec2 mouseVec;
+	SDL_GetMouseState(&mouseVec.x, &mouseVec.y);
+	DeprojectScreenToWorld(mouseVec, glm::vec2(sWidth, sHeight), rayOrigin, rayDirection);
+	auto& dynamicsSystem = Game::instance()->GetDynamicsSystem();
+	if (auto hit = dynamicsSystem.RayCastClosestHit(rayOrigin, rayDirection, 1e10f))
+	{
+		intersectionTransform = hit->first;
+		return std::make_optional(intersectionTransform);
+	}
+	else if (glm::intersectRayPlane(rayOrigin, rayDirection, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f),
+	                                intersectDistance))
+	{
+		intersectionTransform.position = rayOrigin + rayDirection * intersectDistance;
+		intersectionTransform.rotation = glm::mat3(1.0f);
+		return std::make_optional(intersectionTransform);
+	}
+	return std::nullopt;
+}
+
+void Camera::FlyInit()
+{
+	_flyInProgress = false;
+	_flyDist = 0.0f;
+	_flyProgress = 1.0f;
+	_flySpeed = 0.5f;
+	_flyStartAngle = 10.0f;
+	_flyEndAngle = 30.0f;
+	_flyThreshold = 50.0f;
+	_flyFromPos = glm::vec3(0.0f, 0.0f, 0.0f);
+	_flyFromTan = glm::vec3(0.0f, 0.0f, 0.0f);
+	_flyToPos = glm::vec3(0.0f, 0.0f, 0.0f);
+	_flyToNorm = glm::vec3(0.0f, 0.0f, 0.0f);
+	_flyToTan = glm::vec3(0.0f, 0.0f, 0.0f);
+	_flyPrevPos = glm::vec3(0.0f, 0.0f, 0.0f);
+}
+
+glm::vec3 Camera::Fly(glm::vec3& fromPos, glm::vec3& fromTan, glm::vec3& toPos, glm::vec3& toTan, float t)
```

Comment:
You should const qualify the ref arguments that aren't meant to change.

Can you also explain what this function does? It's not very clear from the name and body.

Reply:
deleted and code moved out as used only once.